Assets/Plugins/AudioPluginExample.bundle/Contents/Resources/README.md

```diff
@@ -0,0 +1,81 @@
+# A Revised SDK for Unity Naive Audio Plug-ins
+
+This repository contains a proposal for an updated Unity Native Audio Plugin SDK.
+
+## Why
+
+Working with the original SDK, I found that it is quite convoluted and exposes
+an unnecessary amount of the low-level C API to users. Furthermore, it uses the
+C macro preprocessor extensively which is fragile and cumbersome to debug.
+
+Moreover, the project is somewhat of a mouthful with almost 30 audio effects and
+instruments of varying complexity. A lot of those plug-ins are not
+production-ready as they do memory allocation and mutex locking on the audio
+thread. This includes the convolution reverb. In my opinion, a leaner project
+with a few, simple plug-ins that demonstrate the SDK, is a better approach,
+provided our goal is to teach users how the SDK works.
+
+Also, right now the 'AudioPluginUtil' component is a frankenstein mash-up of
+essential functions related to the SDK and convenience functions for doing
+complex number algebra (why don't we use std::complex?!), Fourier transforms,
+mutex locking (users should probably use std::mutex instead), and other stuff.
+
+Finally, the original SDK needs some love anyway, as the Xcode project no longer
+compiles out-of-the-box (haven't been able to test the VS project).
+
+## Approach
+
+I've created a clean C++17 layer that completely shields the low-level API from
+the user. It declares a handful of interfaces that users implement and then the
+SDK handles all low-level stuff behind the scenes. All of the SDK that users
+need to understand is contained in a single header,
+'UnityAudioPluginInterface.hpp', which is about 100 lines of code. Having a
+clean and self-contained C++ interface simplifies the mental burden that we put
+on our users. Also, getting rid of all the macro-preprocessor non-sense and
+weird callbacks such as 'AudioPluginUtil::RegisterParameter' from the original
+SDK, makes plug-ins easier to maintain and debug.
+
+On top of the C++ interface, I have implemented two simple plug-ins, an
+attenuator and a noise generator. Both are very simple plugins. We could probably
+do with an extra and more complicated plug-in with multiple parameters and more
+advanced DSP.
+
+I have also created a Unity project, an Xcode project (for macOS) and a Visual
+Studio project (for Windows) and tested that the code compiles on both platforms.
+On macOS, I have verified that the plug-ins can load and that they work as expected.
+I haven't been able to do this on Windows as I don't have a Windows machine yet.
+
+NOTE: I haven't made any changes to the low-level C API. (So far!)
+
+## Unclarified Points
+
+1. We need to discuss what version of Unity / Xcode / Visual Studio we want to
+use for the SDK.
+
+2. The way parameters is exposed in the new SDK via a callback function that
+returns a vector of parameter descriptors (in the 'EntryPoint' struct) is not
+as userfriendly as it could be.
+
+3. Is it ok to use the C++ Standard Library (STL)?
```

Comment:
You can definitely use the standard library (not the STL though, that's a historical artifact ;) )

But be careful about enforcing usage of these containers in scenarios where memory allocation might need to be controlled by users

Reply:
Okay, we need to discuss this.